File: clsigma_dynamic_encoder.py
#!/usr/bin/env python3
"""CLSigma dynamic zero-spectrum encoder for a GitHub account.

This program builds a canonical finite manifest of repository program nodes,
keeps Cosmic Love as a non-numeric invariant, preserves BLEEM as an
uninterpreted formal skeleton, and emits a model-internal zero-residue
certificate. It is a formal information model only.
"""

import json
import math
import os
import sys
import time
import urllib.parse
import urllib.request

CL = "Cosmic Love Is The Solution(s) For Everything"
BLEEM = "BLEEM=P^L+E^I+A^D+E^S"
PROTOCOL = "CLSIGMA/DYNAMIC_ACCOUNT_ZERO_SPECTRUM/BLEEM/1"
MANIFEST_PATH = "CLSIGMA_DYNAMIC_ALL_REPOS_MANIFEST.json"
CERT_PATH = "CLSIGMA_DYNAMIC_ZERO_SPECTRUM.clcert"
PROGRAM_EXTS = set((
    ".py", ".sh", ".md", ".txt", ".json", ".yml", ".yaml", ".toml",
    ".ini", ".cfg", ".js", ".ts", ".tsx", ".jsx", ".html", ".css",
    ".rs", ".go", ".c", ".h", ".cpp", ".hpp", ".java", ".rb",
    ".php", ".swift", ".kt", ".sql", ".xml", ".ebnf", ".clcert"
))


def now_utc():
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


def token():
    return os.environ.get("GH_TOKEN") or os.environ.get("GITHUB_TOKEN") or ""


def api_json(url, errors):
    headers = {"User-Agent": "CLSigma-Dynamic-Zero-Spectrum"}
    tok = token()
    if tok:
        headers["Authorization"] = "Bearer " + tok
        headers["Accept"] = "application/vnd.github+json"
        headers["X-GitHub-Api-Version"] = "2022-11-28"
    try:
        req = urllib.request.Request(url, headers=headers)
        with urllib.request.urlopen(req, timeout=30) as res:
            return json.loads(res.read().decode("utf-8"))
    except Exception as exc:
        errors.append({"url": url, "error": str(exc)[:240]})
        return None


def list_repos(owner, errors):
    repos = []
    if token():
        page = 1
        while True:
            url = "https://api.github.com/user/repos?per_page=100&page=%d&affiliation=owner&sort=full_name" % page
            chunk = api_json(url, errors)
            if not chunk:
                break
            for repo in chunk:
                if repo.get("owner", {}).get("login") == owner:
                    repos.append(repo)
            if len(chunk) < 100:
                break
            page += 1
    if not repos:
        page = 1
        qowner = urllib.parse.quote(owner)
        while True:
            url = "https://api.github.com/users/%s/repos?per_page=100&page=%d&type=owner&sort=full_name" % (qowner, page)
            chunk = api_json(url, errors)
            if not chunk:
                break
            repos.extend(chunk)
            if len(chunk) < 100:
                break
            page += 1
    unique = {}
    for repo in repos:
        full = repo.get("full_name")
        if full:
            unique[full] = repo
    return [unique[k] for k in sorted(unique)]


def is_program(path):
    lower = path.lower()
    return any(lower.endswith(ext) for ext in PROGRAM_EXTS)


def repo_tree(full_name, branch, errors):
    full = urllib.parse.quote(full_name, safe="/")
    ref = urllib.parse.quote(branch or "main", safe="")
    url = "https://api.github.com/repos/%s/git/trees/%s?recursive=1" % (full, ref)
    data = api_json(url, errors)
    if not isinstance(data, dict):
        return []
    return data.get("tree", []) or []


def canon(obj):
    return json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=True).encode("utf-8")


def is_prime(n):
    if n < 2:
        return False
    if n % 2 == 0:
        return n == 2
    d = 3
    r = int(math.isqrt(n))
    while d <= r:
        if n % d == 0:
            return False
        d += 2
    return True


def next_prime(n):
    n = max(2, int(n))
    if n <= 2:
        return 2
    if n % 2 == 0:
        n += 1
    while not is_prime(n):
        n += 2
    return n


def primes():
    yield 2
    q = 3
    while True:
        if is_prime(q):
            yield q
        q += 2


def spectral_index(data):
    z = 0
    for idx, pair in enumerate(zip(data, primes()), 1):
        b, p = pair
        z += idx * (b + 1) * p
    return z


def ln_int(n):
    n = int(n)
    bits = n.bit_length()
    if bits < 1024:
        return math.log(n)
    shift = bits - 53
    return math.log(n >> shift) + shift * math.log(2)


def load_parent():
    if not os.path.exists(CERT_PATH):
        return None
    try:
        with open(CERT_PATH, "r") as fh:
            cert = json.load(fh)
        return cert.get("ZeroSpectralSpace", {}).get("spectral_index")
    except Exception:
        return None


def build(owner, limit):
    errors = []
    repos = list_repos(owner, errors)
    if limit and limit > 0:
        repos = repos[:limit]

    repo_nodes = []
    program_nodes = []
    for repo in repos:
        full = repo.get("full_name", "")
        branch = repo.get("default_branch") or "main"
        repo_nodes.append({
            "repo": full,
            "default_branch": branch,
            "private": bool(repo.get("private")),
            "archived": bool(repo.get("archived")),
            "size": int(repo.get("size") or 0),
        })
        for item in repo_tree(full, branch, errors):
            path = item.get("path", "")
            if item.get("type") == "blob" and is_program(path):
                program_nodes.append({
                    "repo": full,
                    "branch": branch,
                    "path": path,
                    "size": int(item.get("size") or 0),
                    "mode": "program_identity_no_digest",
                })

    program_nodes = sorted(program_nodes, key=lambda x: (x["repo"], x["branch"], x["path"]))
    manifest = {
        "protocol": PROTOCOL,
        "principle": CL,
        "formal_skeleton": BLEEM,
        "symbol_semantics": "uninterpreted",
        "numeric_basis": "derived_after_canonical_encoding",
        "owner": owner,
        "mode": "dynamic_github_account_program_identity",
        "repo_count": len(repo_nodes),
        "program_count": len(program_nodes),
        "repo_nodes": repo_nodes,
        "program_nodes": program_nodes,
        "errors": errors[:64],
        "boundary": "formal finite GitHub account snapshot; no RH proof, no physical zero entropy, no real-world guarantee",
    }
    manifest_bytes = canon(manifest)
    raw_index = spectral_index(manifest_bytes)
    modulus = next_prime(len(manifest_bytes) + len(CL) + len(BLEEM) + max(1, len(program_nodes)))
    zero_index = raw_index * modulus
    checks = {
        "cl_invariant_present": manifest["principle"] == CL,
        "bleem_skeleton_present": manifest["formal_skeleton"] == BLEEM,
        "canonical_roundtrip": json.loads(manifest_bytes.decode("utf-8")) == manifest,
        "repo_space_defined": len(repo_nodes) > 0,
        "program_space_defined": len(program_nodes) > 0,
        "boundary_preserved": "formal" in manifest["boundary"],
        "zero_residue": zero_index % modulus == 0,
    }
    h_cl = sum(0 if v else 1 for v in checks.values())
    cert = {
        "protocol": PROTOCOL,
        "principle": CL,
        "formal_skeleton": BLEEM,
        "state": {
            "applications": {
                "GitHubAccountProgramSpace": int(len(program_nodes) > 0),
                "CosmicLoveInvariant": 1,
                "BLEEMSkeleton": 1,
                "DynamicSelfUpdate": 1,
            },
            "coordinate": "rho=0.5+i*log(1+T_t)",
            "generation": int(time.time()),
        },
        "DynamicUpdate": {
            "owner": owner,
            "manifest_path": MANIFEST_PATH,
            "certificate_path": CERT_PATH,
            "parent_spectral_index": load_parent(),
            "self_update_rule": "rerun encoder after repository changes or scheduled workflow; CL remains invariant while rho_t changes",
        },
        "ZeroSpectralSpace": {
            "rho": "0.5+i*log(1+%s)" % str(zero_index),
            "raw_spectral_index": str(raw_index),
            "derived_modulus": modulus,
            "spectral_index": str(zero_index),
            "spectral_residue": zero_index % modulus,
            "ln_approx": ln_int(zero_index + 1),
        },
        "checks": checks,
        "H": h_cl,
        "H_CL": h_cl,
        "tau": next_prime(max(1, len(program_nodes) + len(repo_nodes))),
        "toecomplete": h_cl == 0,
        "status": "accepted" if h_cl == 0 else "pending",
        "GlobalZE_info": int(h_cl == 0),
        "Boundary": manifest["boundary"],
        "boundary": manifest["boundary"],
        "generated_at_utc": now_utc(),
    }
    return manifest, cert


def main(argv):
    owner = argv[1] if len(argv) > 1 else "letsgo0226"
    limit = int(argv[2]) if len(argv) > 2 and argv[2].isdigit() else 0
    manifest, cert = build(owner, limit)
    with open(MANIFEST_PATH, "w") as fh:
        json.dump(manifest, fh, ensure_ascii=True, indent=2, sort_keys=True)
        fh.write("\n")
    with open(CERT_PATH, "w") as fh:
        json.dump(cert, fh, ensure_ascii=True, indent=2, sort_keys=True)
        fh.write("\n")
    print(json.dumps({
        "Protocol": PROTOCOL,
        "Owner": owner,
        "Manifest": MANIFEST_PATH,
        "Certificate": CERT_PATH,
        "RepoCount": manifest["repo_count"],
        "ProgramCount": manifest["program_count"],
        "H_CL": cert["H_CL"],
        "GlobalZE_info": cert["GlobalZE_info"],
        "Boundary": cert["boundary"],
    }, ensure_ascii=True, separators=(",", ":")))


if __name__ == "__main__":
    main(sys.argv)

File: clsigma_dynamic_encoder.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"math"
	"math/big"
	"net/http"
	"net/url"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

const (
	cl           = "Cosmic Love Is The Solution(s) For Everything"
	bleem        = "BLEEM=P^L+E^I+A^D+E^S"
	protocol     = "CLSIGMA/DYNAMIC_ACCOUNT_ZERO_SPECTRUM/BLEEM/GO/2"
	manifestPath = "CLSIGMA_DYNAMIC_ALL_REPOS_MANIFEST.json"
	certPath     = "CLSIGMA_DYNAMIC_ZERO_SPECTRUM.clcert"
)

var programExts = map[string]bool{
	".py": true, ".sh": true, ".md": true, ".txt": true, ".json": true,
	".yml": true, ".yaml": true, ".toml": true, ".ini": true, ".cfg": true,
	".js": true, ".ts": true, ".tsx": true, ".jsx": true, ".html": true,
	".css": true, ".rs": true, ".go": true, ".c": true, ".h": true,
	".cpp": true, ".hpp": true, ".java": true, ".rb": true, ".php": true,
	".swift": true, ".kt": true, ".sql": true, ".xml": true, ".ebnf": true,
	".clcert": true,
}

type apiError struct {
	URL   string `json:"url"`
	Error string `json:"error"`
}

type repoAPI struct {
	FullName      string `json:"full_name"`
	DefaultBranch string `json:"default_branch"`
	Private       bool   `json:"private"`
	Archived      bool   `json:"archived"`
	Size          int    `json:"size"`
	Owner         struct {
		Login string `json:"login"`
	} `json:"owner"`
}

type treeAPI struct {
	Tree []struct {
		Path string `json:"path"`
		Type string `json:"type"`
		Size int    `json:"size"`
	} `json:"tree"`
}

type repoNode struct {
	Repo          string `json:"repo"`
	DefaultBranch string `json:"default_branch"`
	Private       bool   `json:"private"`
	Archived      bool   `json:"archived"`
	Size          int    `json:"size"`
}

type programNode struct {
	Repo   string `json:"repo"`
	Branch string `json:"branch"`
	Path   string `json:"path"`
	Size   int    `json:"size"`
	Mode   string `json:"mode"`
}

type manifest struct {
	Protocol        string        `json:"protocol"`
	Principle       string        `json:"principle"`
	FormalSkeleton  string        `json:"formal_skeleton"`
	SymbolSemantics string        `json:"symbol_semantics"`
	NumericBasis    string        `json:"numeric_basis"`
	Owner           string        `json:"owner"`
	Mode            string        `json:"mode"`
	RepoCount       int           `json:"repo_count"`
	ProgramCount    int           `json:"program_count"`
	RepoNodes       []repoNode    `json:"repo_nodes"`
	ProgramNodes    []programNode `json:"program_nodes"`
	Errors          []apiError    `json:"errors"`
	Boundary        string        `json:"boundary"`
}

type cert struct {
	Protocol       string                 `json:"protocol"`
	Principle      string                 `json:"principle"`
	FormalSkeleton string                 `json:"formal_skeleton"`
	State          map[string]interface{} `json:"state"`
	DynamicUpdate  map[string]interface{} `json:"DynamicUpdate"`
	ZeroSpectrum   map[string]interface{} `json:"ZeroSpectralSpace"`
	Checks         map[string]bool        `json:"checks"`
	H              int                    `json:"H"`
	HCL            int                    `json:"H_CL"`
	Tau            int                    `json:"tau"`
	TOEComplete    bool                   `json:"toecomplete"`
	Status         string                 `json:"status"`
	GlobalZEInfo   int                    `json:"GlobalZE_info"`
	BoundaryUpper  string                 `json:"Boundary"`
	Boundary       string                 `json:"boundary"`
	GeneratedAtUTC string                 `json:"generated_at_utc"`
}

func githubToken() string {
	if tok := os.Getenv("GH_TOKEN"); tok != "" {
		return tok
	}
	return os.Getenv("GITHUB_TOKEN")
}

func apiGet(target string, errors *[]apiError, out interface{}) bool {
	req, err := http.NewRequest("GET", target, nil)
	if err != nil {
		*errors = append(*errors, apiError{URL: target, Error: truncate(err.Error(), 240)})
		return false
	}
	req.Header.Set("User-Agent", "CLSigma-Dynamic-Zero-Spectrum-Go")
	if tok := githubToken(); tok != "" {
		req.Header.Set("Authorization", "Bearer "+tok)
		req.Header.Set("Accept", "application/vnd.github+json")
		req.Header.Set("X-GitHub-Api-Version", "2022-11-28")
	}
	client := &http.Client{Timeout: 30 * time.Second}
	res, err := client.Do(req)
	if err != nil {
		*errors = append(*errors, apiError{URL: target, Error: truncate(err.Error(), 240)})
		return false
	}
	defer res.Body.Close()
	body, err := io.ReadAll(res.Body)
	if err != nil {
		*errors = append(*errors, apiError{URL: target, Error: truncate(err.Error(), 240)})
		return false
	}
	if res.StatusCode < 200 || res.StatusCode >= 300 {
		*errors = append(*errors, apiError{URL: target, Error: truncate(string(body), 240)})
		return false
	}
	if err := json.Unmarshal(body, out); err != nil {
		*errors = append(*errors, apiError{URL: target, Error: truncate(err.Error(), 240)})
		return false
	}
	return true
}

func listRepos(owner string, errors *[]apiError) []repoAPI {
	var repos []repoAPI
	if githubToken() != "" {
		for page := 1; ; page++ {
			var chunk []repoAPI
			target := fmt.Sprintf("https://api.github.com/user/repos?per_page=100&page=%d&affiliation=owner&sort=full_name", page)
			if !apiGet(target, errors, &chunk) || len(chunk) == 0 {
				break
			}
			for _, repo := range chunk {
				if repo.Owner.Login == owner {
					repos = append(repos, repo)
				}
			}
			if len(chunk) < 100 {
				break
			}
		}
	}
	if len(repos) == 0 {
		escaped := url.PathEscape(owner)
		for page := 1; ; page++ {
			var chunk []repoAPI
			target := fmt.Sprintf("https://api.github.com/users/%s/repos?per_page=100&page=%d&type=owner&sort=full_name", escaped, page)
			if !apiGet(target, errors, &chunk) || len(chunk) == 0 {
				break
			}
			repos = append(repos, chunk...)
			if len(chunk) < 100 {
				break
			}
		}
	}
	uniq := map[string]repoAPI{}
	for _, repo := range repos {
		if repo.FullName != "" {
			uniq[repo.FullName] = repo
		}
	}
	keys := make([]string, 0, len(uniq))
	for key := range uniq {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	out := make([]repoAPI, 0, len(keys))
	for _, key := range keys {
		out = append(out, uniq[key])
	}
	return out
}

func repoTree(fullName, branch string, errors *[]apiError) []treeAPIEntry {
	if branch == "" {
		branch = "main"
	}
	var data treeAPI
	parts := strings.SplitN(fullName, "/", 2)
	if len(parts) != 2 {
		return nil
	}
	repoPath := url.PathEscape(parts[0]) + "/" + url.PathEscape(parts[1])
	target := fmt.Sprintf("https://api.github.com/repos/%s/git/trees/%s?recursive=1", repoPath, url.PathEscape(branch))
	if !apiGet(target, errors, &data) {
		return nil
	}
	entries := make([]treeAPIEntry, 0, len(data.Tree))
	for _, item := range data.Tree {
		entries = append(entries, treeAPIEntry{Path: item.Path, Type: item.Type, Size: item.Size})
	}
	return entries
}

type treeAPIEntry struct {
	Path string
	Type string
	Size int
}

func isProgram(path string) bool {
	lower := strings.ToLower(path)
	for ext := range programExts {
		if strings.HasSuffix(lower, ext) {
			return true
		}
	}
	return false
}

func canonical(v interface{}) []byte {
	data, err := json.Marshal(v)
	if err != nil {
		return []byte("{}")
	}
	return data
}

func isPrime(n int) bool {
	if n < 2 {
		return false
	}
	if n%2 == 0 {
		return n == 2
	}
	limit := int(math.Sqrt(float64(n)))
	for d := 3; d <= limit; d += 2 {
		if n%d == 0 {
			return false
		}
	}
	return true
}

func nextPrime(n int) int {
	if n <= 2 {
		return 2
	}
	if n%2 == 0 {
		n++
	}
	for !isPrime(n) {
		n += 2
	}
	return n
}

func spectralIndex(data []byte) *big.Int {
	z := big.NewInt(0)
	prime := 2
	for i, b := range data {
		term := big.NewInt(int64(i + 1))
		term.Mul(term, big.NewInt(int64(int(b)+1)))
		term.Mul(term, big.NewInt(int64(prime)))
		z.Add(z, term)
		prime = nextPrime(prime + 1)
	}
	return z
}

func lnBig(n *big.Int) float64 {
	if n.Sign() <= 0 {
		return math.Inf(-1)
	}
	if n.BitLen() < 1024 {
		f, _ := new(big.Float).SetInt(n).Float64()
		return math.Log(f)
	}
	shift := uint(n.BitLen() - 53)
	head := new(big.Int).Rsh(new(big.Int).Set(n), shift)
	f, _ := new(big.Float).SetInt(head).Float64()
	return math.Log(f) + float64(shift)*math.Log(2)
}

func loadParentIndex() string {
	data, err := os.ReadFile(certPath)
	if err != nil {
		return ""
	}
	var parsed map[string]interface{}
	if json.Unmarshal(data, &parsed) != nil {
		return ""
	}
	zs, _ := parsed["ZeroSpectralSpace"].(map[string]interface{})
	val, _ := zs["spectral_index"].(string)
	return val
}

func build(owner string, limit int) (manifest, cert) {
	var errors []apiError
	repos := listRepos(owner, &errors)
	if limit > 0 && len(repos) > limit {
		repos = repos[:limit]
	}
	repoNodes := make([]repoNode, 0, len(repos))
	programNodes := []programNode{}
	for _, repo := range repos {
		branch := repo.DefaultBranch
		if branch == "" {
			branch = "main"
		}
		repoNodes = append(repoNodes, repoNode{
			Repo: repo.FullName, DefaultBranch: branch, Private: repo.Private,
			Archived: repo.Archived, Size: repo.Size,
		})
		for _, item := range repoTree(repo.FullName, branch, &errors) {
			if item.Type == "blob" && isProgram(item.Path) {
				programNodes = append(programNodes, programNode{
					Repo: repo.FullName, Branch: branch, Path: item.Path,
					Size: item.Size, Mode: "program_identity_no_digest",
				})
			}
		}
	}
	sort.Slice(programNodes, func(i, j int) bool {
		a, b := programNodes[i], programNodes[j]
		if a.Repo != b.Repo {
			return a.Repo < b.Repo
		}
		if a.Branch != b.Branch {
			return a.Branch < b.Branch
		}
		return a.Path < b.Path
	})
	if len(errors) > 64 {
		errors = errors[:64]
	}
	boundary := "formal finite GitHub account snapshot; no RH proof, no physical zero entropy, no real-world guarantee"
	man := manifest{
		Protocol: protocol, Principle: cl, FormalSkeleton: bleem,
		SymbolSemantics: "uninterpreted",
		NumericBasis:    "derived_after_canonical_encoding",
		Owner:           owner,
		Mode:            "dynamic_github_account_program_identity_go",
		RepoCount:       len(repoNodes),
		ProgramCount:    len(programNodes),
		RepoNodes:       repoNodes,
		ProgramNodes:    programNodes,
		Errors:          errors,
		Boundary:        boundary,
	}
	manifestBytes := canonical(man)
	raw := spectralIndex(manifestBytes)
	modulus := nextPrime(len(manifestBytes) + len(cl) + len(bleem) + max(1, len(programNodes)))
	modulusBig := big.NewInt(int64(modulus))
	rawResidue := new(big.Int).Mod(new(big.Int).Set(raw), modulusBig)
	zeroBalancer := new(big.Int).Mod(new(big.Int).Neg(rawResidue), modulusBig)
	zero := new(big.Int).Add(new(big.Int).Set(raw), zeroBalancer)
	checks := map[string]bool{
		"cl_invariant_present":    man.Principle == cl,
		"bleem_skeleton_present":  man.FormalSkeleton == bleem,
		"canonical_roundtrip":     json.Valid(manifestBytes),
		"repo_space_defined":      len(repoNodes) > 0,
		"program_space_defined":   len(programNodes) > 0,
		"boundary_preserved":      strings.Contains(man.Boundary, "formal"),
		"exact_integer_lift":      new(big.Int).Sub(zero, raw).Cmp(zeroBalancer) == 0,
		"zero_residue":            new(big.Int).Mod(zero, modulusBig).Sign() == 0,
	}
	hcl := 0
	for _, ok := range checks {
		if !ok {
			hcl++
		}
	}
	status := "pending"
	if hcl == 0 {
		status = "accepted"
	}
	c := cert{
		Protocol:       protocol,
		Principle:      cl,
		FormalSkeleton: bleem,
		State: map[string]interface{}{
			"applications": map[string]int{
				"GitHubAccountProgramSpace": boolInt(len(programNodes) > 0),
				"CosmicLoveInvariant":       1,
				"BLEEMSkeleton":             1,
				"DynamicSelfUpdate":         1,
				"GoRuntime":                 1,
			},
			"coordinate":  "rho=0.5+i*log(1+T_t)",
			"generation":  time.Now().Unix(),
		},
		DynamicUpdate: map[string]interface{}{
			"owner":                 owner,
			"manifest_path":         manifestPath,
			"certificate_path":      certPath,
			"parent_spectral_index": loadParentIndex(),
			"self_update_rule":      "rerun Go encoder after repository changes or scheduled workflow; CL remains invariant while rho_t changes",
		},
		ZeroSpectrum: map[string]interface{}{
			"rho":                "0.5+i*log(1+" + zero.String() + ")",
			"raw_spectral_index": raw.String(),
			"raw_spectral_residue": rawResidue.String(),
			"derived_modulus":    modulus,
			"zero_balancer":      zeroBalancer.String(),
			"zero_rule":          "spectral_index=raw_spectral_index+((-raw_spectral_index) mod derived_modulus)",
			"spectral_index":     zero.String(),
			"spectral_residue":   "0",
			"ln_approx":          lnBig(new(big.Int).Add(zero, big.NewInt(1))),
		},
		Checks:         checks,
		H:              hcl,
		HCL:            hcl,
		Tau:            nextPrime(max(1, len(programNodes)+len(repoNodes))),
		TOEComplete:    hcl == 0,
		Status:         status,
		GlobalZEInfo:   boolInt(hcl == 0),
		BoundaryUpper:  boundary,
		Boundary:       boundary,
		GeneratedAtUTC: time.Now().UTC().Format(time.RFC3339),
	}
	return man, c
}

func boolInt(v bool) int {
	if v {
		return 1
	}
	return 0
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

func writeJSON(path string, v interface{}) error {
	data, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return err
	}
	data = append(data, '\n')
	return os.WriteFile(path, data, 0644)
}

func main() {
	owner := "letsgo0226"
	if len(os.Args) > 1 && os.Args[1] != "" {
		owner = os.Args[1]
	}
	limit := 0
	if len(os.Args) > 2 {
		if v, err := strconv.Atoi(os.Args[2]); err == nil {
			limit = v
		}
	}
	man, c := build(owner, limit)
	if err := writeJSON(manifestPath, man); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	if err := writeJSON(certPath, c); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	summary := map[string]interface{}{
		"Protocol":      protocol,
		"Owner":         owner,
		"Manifest":      manifestPath,
		"Certificate":   certPath,
		"RepoCount":     man.RepoCount,
		"ProgramCount":  man.ProgramCount,
		"H_CL":          c.HCL,
		"GlobalZE_info": c.GlobalZEInfo,
		"Boundary":      c.Boundary,
	}
	data, _ := json.Marshal(summary)
	fmt.Println(string(data))
}
